geomstats/geometry/stratified/wald_space.py

```diff
@@ -47,949 +15,973 @@ class :class:`spd.SPDMatrices`.
     https://doi.org/10.1007/978-3-030-80209-7_76.
 """
 
-import functools
-import itertools
+from abc import ABC
 
 import geomstats.backend as gs
-import geomstats.geometry.spd_matrices as spd
-from geomstats.geometry.stratified.point_set import Point, PointSet, _vectorize_point
-
+from geomstats.geometry.hermitian_matrices import powermh
+from geomstats.geometry.matrices import Matrices
+from geomstats.geometry.spd_matrices import SPDMatrices
+from geomstats.geometry.stratified.point_set import (
+    Point,
+    PointCollection,
+    PointSet,
+    PointSetMetric,
+)
+from geomstats.geometry.stratified.trees import (
+    ForestTopology,
+    Split,
+    delete_splits,
+    generate_splits,
+)
+from geomstats.geometry.stratified.vectorization import (
+    _manipulate_output,
+    broadcast_lists,
+    vectorize_point,
+)
+from geomstats.numerics.optimizers import ScipyMinimize
+
+
+def _manipulate_input_with_array(arg):
+    if gs.is_array(arg):
+        if arg.ndim > 2:
+            return arg, False
+        return gs.expand_dims(arg, axis=0), True
+
+    if not isinstance(arg, (list, tuple)):
+        return [arg], True
+
+    return arg, False
+
+
+def _manipulate_output_wald(out, to_list):
+    return _manipulate_output(out, to_list, manipulate_output_iterable=WaldCollection)
+
+
+def make_splits(n_labels):
+    """Generate all possible splits of a collection."""
+    if n_labels <= 1:
+        raise ValueError("`n_labels` must be greater than 1.")
+    if n_labels == 2:
+        yield Split(part1=[0], part2=[1])
+    else:
```

Comment:
is else needed after yield?

Reply:
Yes, otherwise also runs code after else.